### PR TITLE
Optimize Match-on-metadata

### PR DESCRIPTION
**Pull request by Forgind**
Fixes [AB#1261123](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1261123) (feedback ticket)

### Context
Checking metadata for matches was taking (#fragments in item) * (#fragments in item to remove) * (#metadata). For many users, this was extremely slow.

### Changes Made
Put fragments into a custom MetadataSet object before checking for a match. This takes it from O(n * r * m) to O(m * (n + r)). Since m is normally small, this is a substantial improvement.

### Testing
Time on a single operation went from ~180 seconds to ~150 ms for the penultimate version. It also passes @cdmihai's extensive unit tests.*

### Notes
Is there any reason to keep the error for referencing multiple items? As a user, I would expect not-match-on-metadata to ignore metadata and match-on-metadata to ignore the item.
Is it ok to rely on IEnumerable returning element in a consistent order?
Should we keep returning no-match if all the metadata are empty on both items? It makes more sense to me to have that be a match, and it would simplify the code slightly. If it makes more sense to others, now would be the time to do it, since the code isn't that old.

\* Except one on paths. Not sure why that's failing, since it's passing in VS, just not from the command line. Will look into it more.

### CODE REVIEW

**Comment by BenVillalobos** [comment]
I believe if you do `AB/<feedback ticket number>` then a tool that kevin pilch made will auto-link to the ticket

File: src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupIntrinsicTask.cs [hunk 1]
```diff
@@ -249,7 +249,18 @@
             }
             else
             {
-                itemsToRemove = FindItemsUsingMatchOnMetadata(group, child, bucket, matchOnMetadata, matchingOptions);
+                ImmutableList<string> metadataList = matchOnMetadata.ToImmutableList();
```
Comment (cdmihai) [inline src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupIntrinsicTask.cs:252]: I'd keep this as a separate method to have ExecuteRemove read nicer by only contain top level intent and delegating more nitty gritty implementation details to helper methods. Like the other if branch.

File: src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs [hunk 2]
```diff
@@ -55,13 +56,23 @@ protected override ImmutableList<I> SelectItems(ImmutableList<ItemData>.Builder
...
+                    if (_matchOnMetadata.IsEmpty ? _itemSpec.MatchesItem(item.Item) : MatchesItemOnMetadata(item.Item))
                         items.Add(item.Item);
                 }
 
                 return items.ToImmutableList();
             }
 
+            private bool MatchesItemOnMetadata(I item)
+            {
+                if (metadataSet == null)
```
Comment (cdmihai) [inline src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs:68]: I'd keep MetadataSet as a hidden implementation detail in ItemSpec. The ItemSpec users do not need to know about it.

File: src/Build/Evaluation/ItemSpec.cs [hunk 3]
```diff
@@ -504,4 +459,78 @@ public GlobFragment(string textFragment, string projectDirectory)
...
+    {
+        private Dictionary<string, MetadataSet<P, I>> children;
+        Func<string, string> normalize;
+
+        internal MetadataSet(MatchOnMetadataOptions options, ImmutableList<string> metadata, ItemSpec<P, I> itemSpec)
+        {
+            StringComparer comparer = options == MatchOnMetadataOptions.CaseSensitive ? StringComparer.Ordinal : StringComparer.OrdinalIgnoreCase;
+            children = new Dictionary<string, MetadataSet<P, I>>(comparer);
+            normalize = options == MatchOnMetadataOptions.PathLike ? p => FileUtilities.NormalizePathForComparisonNoThrow(p, Environment.CurrentDirectory) : p => p;
+            foreach (ItemSpec<P, I>.ItemExpressionFragment frag in itemSpec.Fragments)
```
Comment (cdmihai) [inline src/Build/Evaluation/ItemSpec.cs:473]: How does this type check, isn't ItemSpec.Fragments a list of the base class ItemSpecFragment? (`List<ItemSpecFragment>`)

File: src/Build/Evaluation/ItemSpec.cs [hunk 4]
```diff
@@ -504,4 +459,78 @@ public GlobFragment(string textFragment, string projectDirectory)
...
+
+        internal bool Contains(IEnumerable<string> metadata)
+        {
+            bool nonEmptyFound = false;
+            MetadataSet<P, I> curr = this;
+            foreach (string m in metadata)
+            {
+                if (!String.IsNullOrEmpty(m))
+                {
+                    nonEmptyFound = true;
```
Comment (ladipro) [inline src/Build/Evaluation/ItemSpec.cs:514]: Why isn't this `return false`? The previous code was checking if all metadata is defined *and* their value matches. I suggest adding test coverage for the case where `m` is empty.

**Review by ladipro — COMMENTED** [review]
Neat! I have left comments - mostly nits, but the logic in `Contains` appears to be incorrect, please add a test case exercising all code paths there.

File: src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs [hunk 5]
```diff
@@ -55,13 +56,23 @@ protected override ImmutableList<I> SelectItems(ImmutableList<ItemData>.Builder
...
+                {
+                    metadataSet = new MetadataSet<P, I>(_matchOnMetadataOptions, _matchOnMetadata, _itemSpec);
```
[the remark this answers is not in the document]
Comment (Forgind) [reply src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs:70]: I can put it in the constructor—shouldn't affect perf, assuming operations are always executed.

File: src/Build/Evaluation/ItemSpec.cs (hunk 4 above)
Comment (Forgind) [reply src/Build/Evaluation/ItemSpec.cs:514]: From https://github.com/dotnet/msbuild/pull/4997:
`if a metadata name doesn't exist in both items, then it's considered a match. Special case: if none of the metadata exist in both items, then it should not match.`

But you're right—the code I'm replacing checks for string.Empty and automatically returns false. Given our conversation in the PR review meeting, I'll change to that. Thanks!

File: src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs (hunk 2 above)
Comment (Forgind) [reply src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs:68]: They don't have to know about it anyway. The methods that call it in ItemGroupIntrinsicTask and LazyItemEvaluator.RemoveOperation are both private. I count that as close enough.

File: src/Build/Evaluation/ItemSpec.cs (hunk 3 above)
Comment (Forgind) [reply src/Build/Evaluation/ItemSpec.cs:473]: Good question. It might be casting it automatically when it sees that, but I'm not sure. I don't think we should care about other kinds of fragments, since I don't think they can have metadata, so I think this is ok, but that was a lot of "I think"s.

File: src/Build.UnitTests/BackEnd/IntrinsicTask_Tests.cs [hunk 6]
```diff
@@ -1988,63 +1988,6 @@ public void KeepWithItemReferenceOnNonmatchingMetadata()
             items.ElementAt(3).GetMetadataValue("d").ShouldBe("d");
         }
 
-        [Fact]
-        public void FailWithMatchingMultipleMetadata()
```
Comment (cdmihai) [inline src/Build.UnitTests/BackEnd/IntrinsicTask_Tests.cs:1992]: Don't delete the tests, instead replace the exception assertion with an assertion on I2's contents

File: src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs [hunk 7]
```diff
@@ -2418,8 +2418,9 @@ public void FailWithMatchingMultipleMetadata()
 
                 <I2 Remove='@(I1)' MatchOnMetadata='M1;M2'/>");
 
-            Should.Throw<InvalidProjectFileException>(() => ObjectModelHelpers.CreateInMemoryProject(content))
-                .HelpKeyword.ShouldBe("MSBuild.OM_MatchOnMetadataIsRestrictedToOnlyOneReferencedItem");
+            Project project = ObjectModelHelpers.CreateInMemoryProject(content);
+            IEnumerable<ProjectItem> items = project.ItemsIgnoringCondition.Where(i => i.ItemType.Equals("I2"));
+            items.Count().ShouldBe(3);
```
Comment (cdmihai) [inline src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs:2423]: For both of the tests, assert that the right items are left in I2.

File: src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs (hunk 2 above)
Comment (cdmihai) [reply src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs]: By users I meant Itemspec API users. When somebody tries to understand the code in ItemGroupIntrinsicTask  or in LazyItemEvaluator.RemoveOperation they'll get hit with two concepts they have to understand: Itemspec and MetadataSet. If you hide MetadataSet inside Itemspec it becomes an implementation detail that people don't have to care about unless they're interested in how Itemspec works.

File: src/Build/Evaluation/ItemSpec.cs [hunk 8]
```diff
@@ -504,4 +458,74 @@ public GlobFragment(string textFragment, string projectDirectory)
             && TextFragment[2] == '*'
             && FileUtilities.IsAnySlash(TextFragment[3]);
     }
+
+    internal sealed class MetadataSet<P, I> where P : class, IProperty where I : class, IItem, IMetadataTable
+    {
+        private readonly Dictionary<string, MetadataSet<P, I>> _children;
```
Comment (cdmihai) [inline src/Build/Evaluation/ItemSpec.cs:464]: Why use this convoluted data structure instead of dumping all metadata values from the Itemspec into a flat set of normalized metadata values?

**Review by BenVillalobos — COMMENTED** [review]
> Should we keep returning no-match if all the metadata are empty on both items?

I can theoretically see a use case for "Match on everything that has no metadata." Maybe we can use a special keyword/symbol here to represent that? Might be best to file that as an issue and see if that gets traction.

File: src/Build.UnitTests/BackEnd/IntrinsicTask_Tests.cs [hunk 9]
```diff
@@ -2010,12 +2010,16 @@ public void FailWithMatchingMultipleMetadata()
                     </Target></Project>");
             IntrinsicTask task = CreateIntrinsicTask(content);
             Lookup lookup = LookupHelpers.CreateEmptyLookup();
-            Assert.ThrowsAny<InvalidProjectFileException>(() => ExecuteTask(task, lookup))
-                .HelpKeyword.ShouldBe("MSBuild.OM_MatchOnMetadataIsRestrictedToOnlyOneReferencedItem");
+            ExecuteTask(task, lookup);
+            ICollection<ProjectItemInstance> items = lookup.GetItems("I2");
+            items.Count().ShouldBe(3);
+            items.ElementAt(0).EvaluatedInclude.ShouldBe("a2");
```
[the remark this answers is not in the document]
Comment (Forgind) [reply src/Build.UnitTests/BackEnd/IntrinsicTask_Tests.cs:2016]: MatchOnMetadata allows three options: case sensitive, case insensitive, and pathlike. The third is most similar to what you were thinking, I think, with defaulting to case insensitive regardless of OS.

File: src/Build.UnitTests/BackEnd/IntrinsicTask_Tests.cs (hunk 6 above)
Comment (Forgind) [reply src/Build.UnitTests/BackEnd/IntrinsicTask_Tests.cs:1992]: Changed it to "RemoveWithMatchingMultipleMetdata." I did enable matching multiple.

File: src/Build/Evaluation/ItemSpec.cs (hunk 8 above)
Comment (BenVillalobos) [reply src/Build/Evaluation/ItemSpec.cs:464]: Still not 100% sure I agree with the data structure, but I want to clarify my understanding from the PR review today.

A flat set of normalized metadata values would prevent us from being able to determine exactly _which_ metadata values from _which_ items were matched.

We'd need to do this to enable the logic in a test like `RemoveWithMultipleItemReferenceOnMatchingMetadata`.
```xml
<I3 Remove='@(I1);@(I2)' MatchOnMetadata='M1' />
```
This remove attribute effectively translates to:
Remove items in I3 where:
- I1 **OR** I2 match on the following metadata:
   - M1

I believe this also enables something like:

```xml
<I3 Remove='@(I1);@(I2)' MatchOnMetadata='M1;M2' />
```
Where this remove attribute translates to:
Remove items in I3 where:
- I1 **OR** I2 match on the following metadata:
   - M1 **AND** M2

@Forgind is this understanding correct? If so I think there should be a test for multiple items and multiple metadata.

Comment (Forgind) [reply src/Build/Evaluation/ItemSpec.cs:464]: Wrote detailed comment for MetadataTrie class. Will push soon.

File: src/Build/Evaluation/ItemSpec.cs (hunk 3 above)
Comment (Forgind) [reply src/Build/Evaluation/ItemSpec.cs:509]: As far as whether this is actually a safe, I should have mentioned that the type is checked before the MetadataSet is constructed, so it should always be correct.

File: src/Build/Evaluation/ItemSpec.cs [hunk 10]
```diff
@@ -504,4 +458,111 @@ public GlobFragment(string textFragment, string projectDirectory)
...
+    /// NotTargetFramework netcoreapp3.1.
+    /// 
+    /// Implementing this as a list of sets where each metadatum key has its own set also would not work
+    /// because different items could match on different metadata, and we want to check to see if any
+    /// single item matches on all the metadata. As an example, consider this scenario:
+    /// Item Baby has metadata GoodAt="eating" BadAt="talking" OkAt="sleeping"
+    /// Item Child has metadata GoodAt="sleeping" BadAt="eating" OkAt="talking"
+    /// Item Adolescent has metadata GoodAt="talking" BadAt="sleeping" OkAt="eating"
+    /// Specifying these three metadata:
+    /// Item Forgind with metadata GoodAt="sleeping" BadAt="talking" OkAt="eating"
```
Comment (ladipro) [inline src/Build/Evaluation/ItemSpec.cs:476]: 😄

**Review by ladipro — COMMENTED** [review]
Apologies for getting back to this after so many days.

Playing devil's advocate, imagine that instead of the Trie, we use a simple `HashSet<string>` where the keys are a concatenation of the respective metadata value with a delimiter & proper escaping to make sure that matching doesn't have false positives. What would be the advantages and disadvantages of this approach?

Using your example from the code comment, the HashSet would contain something like:

`"eating|talking|sleeping"`
`"sleeping|eating|talking"`
`"talking|sleeping|eating"`
 
so it wouldn't match the Forgind item with the "signature" of `"sleeping|talking|eating"`.

File: src/Build/Evaluation/ItemSpec.cs (hunk 10 above)
Comment (Forgind) [reply src/Build/Evaluation/ItemSpec.cs:476]: You clearly get the advantage (disadvantage? 😋) of a simpler data structure, but the logic of how to construct the key is more complicated, which partially balances it out. Time to hash a particular string is officially constant but in practice scales according to the length of the string (nlg(n) and assumes the string is short, if I remember correctly), so I don't think the total hashing time would be less. We would win on lookups in the dictionary, but that isn't really relevant. We would also have to factor in time-to-construct the string, which would also always create a new string unless there happens to be exactly one metadatum, and we put in an optimization to not add a character to the end of that. (The argument against an optimization is that we would have to find out that we don't need the extra character, which would either require counting the length of an enumerable (fast, but still theoretically O(n)) or creating something like a StringBuilder, adding it to that StringBuilder, and having a check that the SB has only had one thing added to it before stringifying it. Admittedly, that part should be better with StringTools, so this argument is pretty weak.)

The biggest disadvantage in my mind is with reusing strings, since even if we concatenated them all in one step, we'd still make a new string for each item, and if there are several metadata or the metadata are large (noting that these are metadata values, not metadata keys), it could end up consuming a lot of memory. This is a case in which StringTools would struggle because each of the metadata values, once concatenated to others, could be unique, in which case we'd make a large number of unique large strings. Even matching on two metadata, if the values are large, say, "description" and "path" (with long paths enabled), these could be hundreds or even thousands of characters. With a metadata trie, we don't make any new strings. With a single hash set, we would.

**Comment by Forgind** [comment]
You clearly get the advantage (disadvantage? 😋) of a simpler data structure, but the logic of how to construct the key is more complicated, which partially balances it out. Time to hash a particular string is officially constant but in practice scales according to the length of the string (nlg(n) and assumes the string is short, if I remember correctly), so I don't think the total hashing time would be less. We would win on lookups in the dictionary, but that isn't really relevant. We would also have to factor in time-to-construct the string, which would also always create a new string unless there happens to be exactly one metadatum, and we put in an optimization to not add a character to the end of that. (The argument against an optimization is that we would have to find out that we don't need the extra character, which would either require counting the length of an enumerable (fast, but still theoretically O(n)) or creating something like a StringBuilder, adding it to that StringBuilder, and having a check that the SB has only had one thing added to it before stringifying it. Admittedly, that part should be better with StringTools, so this argument is pretty weak.)

The biggest disadvantage in my mind is with reusing strings, since even if we concatenated them all in one step, we'd still make a new string for each item, and if there are several metadata or the metadata are large (noting that these are metadata values, not metadata keys), it could end up consuming a lot of memory. This is a case in which StringTools would struggle because each of the metadata values, once concatenated to others, could be unique, in which case we'd make a large number of unique large strings. Even matching on two metadata, if the values are large, say, "description" and "path" (with long paths enabled), these could be hundreds or even thousands of characters. With a metadata trie, we don't make any new strings. With a single hash set, we would.

**Review by cdmihai — APPROVED** [review]

**Review by KirillOsenkov — APPROVED** [review]

**Comment by ladipro** [comment]
Thank you for the detailed analysis!

> With a metadata trie, we don't make any new strings.

Actually we do if `MatchOnMetadataOptions.PathLike` is used, though I agree that usually the footprint of the Trie would be smaller than that of a flat set. As long as the Trie is not significantly less efficient in the worst case (all values distinct and path-normalized) I think that a slightly more complex datastructure is justified. Thanks again for taking care of this one!

**Review by ladipro — APPROVED** [review]
